### PR TITLE
Basic AI

### PR DESCRIPTION
**Pull request by JVH-Software**

### CODE REVIEW

File: Assets/Scripts/EnemyPack.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    new void Start()
+    {
+        base.Start();
+    }
+
+    void Update()
+    {
+        // Movement
+        List<Node> path;
+        Pathfinding.FindPath(new Vector2Int(Mathf.RoundToInt(transform.position.x -0.5f), Mathf.RoundToInt(transform.position.y - 0.5f)),
```
Comment (davidhampson) [inline Assets/Scripts/EnemyPack.cs:22]: Comments here please <3

File: Assets/Scripts/Node.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        Position = new Vector2Int(x, y);
+        this.Passable = passable;
+    }
+
+    public override string ToString()
+    {
+        return "x: " + Position.x + " y: " + Position.y + " passable: " + Passable;
+    }
+
+    public override bool Equals(object obj)
```
Comment (SamDelaney) [inline Assets/Scripts/Node.cs:30]: Move in front of the other Equals method since it references it.

File: Assets/Scripts/Node.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+/// </summary>
+public class Node
+{
+    public bool Passable { get; set; }
+    public int gCost { get; set; }
+    public int hCost { get; set; }
+    public Vector2Int Position { get; set; }
+    public Node Parent { get; set; }
+
+    public int fCost
```
Comment (SamDelaney) [inline Assets/Scripts/Node.cs:14]: Comments!

File: Assets/Scripts/Pathfinding.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    /// <summary>
+    /// Looks for a path
+    /// </summary>
+    /// <param name="start">The starting position</param>
+    /// <param name="target">The destination the path should lead to</param>
+    /// <param name="path">The path created</param>
+    public static void FindPath(Vector2Int start, Vector2Int target, Tilemap tilemap, out List<Node> path)
+    {
+        Debug.Log("looking for path");
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs]: Don't push debug logs. We will forget about them and publish them.

File: Assets/Scripts/Pathfinding.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        while (openSet.Count > 0)
+        {
+            Node currentNode = openSet[0];
+
+            openSet.Remove(currentNode);
+            closedSet.Add(currentNode);
+
+            if (currentNode.Equals(targetNode))
+            {
+                path = RetracePath(startNode, currentNode);
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs:39]: explain this

File: Assets/Scripts/Pathfinding.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    /// <param name="target">The destination the path should lead to</param>
+    /// <param name="path">The path created</param>
+    public static void FindPath(Vector2Int start, Vector2Int target, Tilemap tilemap, out List<Node> path)
+    {
+        Debug.Log("looking for path");
+        path = new List<Node>();
+        Node startNode = new Node(start.x, start.y, true);
+        Node targetNode = new Node(target.x, target.y, true);
+
+        List<Node> openSet = new List<Node>();
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs:25]: explain these

File: Assets/Scripts/Pathfinding.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            closedSet.Add(currentNode);
+
+            if (currentNode.Equals(targetNode))
+            {
+                path = RetracePath(startNode, currentNode);
+                return;
+            }
+
+            foreach (Node neighbour in GetNeighbours(currentNode, tilemap))
+            {
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs:44]: comment this entire block

File: Assets/Scripts/Pathfinding.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        List<Node> path = new List<Node>();
+        Node currentNode = endNote;
+
+        while (!currentNode.Equals(startNode))
+        {
+            path.Add(currentNode);
+            currentNode = currentNode.Parent;
+        }
+        Debug.Log("path found, steps: " + path.Count);
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs]: debug log

File: Assets/Scripts/Pathfinding.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+        path.Reverse();
+        return path;
+    }
+
+    private static int GetDistance(Node a, Node b)
+    {
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs:83]: explain this. particularly the random multiplication by 10 and 14

File: Assets/Scripts/Pathfinding.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        int dstX = Mathf.Abs(a.Position.x - b.Position.x);
+        int dstY = Mathf.Abs(a.Position.y - b.Position.y);
+
+        if (dstX > dstY)
+            return 14 * dstY + 10 * (dstX - dstY);
+
+        return 14 * dstX + 10 * (dstY - dstX);
+    }
+
+    private static List<Node> GetNeighbours(Node node, Tilemap tilemap)
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs:93]: comment this method

File: Assets/Scripts/Pathfinding.cs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+        for (int x = -1; x <= 1; x++)
+        {
+            for (int y = -1; y <= 1; y++)
+            {
+                if (x == 0 && y == 0)
+                    continue;
+
+                // No cutting corners
+                if (Mathf.Abs(x) + Mathf.Abs(y) == 2)
```
Comment (SamDelaney) [inline Assets/Scripts/Pathfinding.cs:105]: add a little more detail to how this works

**Review by SamDelaney — APPROVED** [review]

**Comment by SamDelaney** [comment]
[Pack AI](https://trello.com/c/MB8GtmeQ/3-pack-ai)

**Pull request merged**